### PR TITLE
Add methods to NodeAPI for signmessage and checkmessage [issue #109]

### PR DESCRIPTION
**Pull request by lnd3v**
Early (partial) commit.

This commit augments the Breez SDK with message signing and verification facilities. Changes span across the Node API interface, its implementation (Greenlight), BreezServices and the UniFFI binding, with amendments in the FFI definition.

- Added sign_message() and check_message() methods to the NodeAPI trait and Greenlight.
- Added async calls for these methods in BreezServices.
- Exposed said calls via the UniFFI binding for FFI.
- Updated the FFI definition to mirror these additions.

Todo:

[ x] formatting
[ x] tests
[ ] etc.

### CODE REVIEW

File: libs/sdk-core/src/breez_services.rs [hunk 1]
```diff
@@ -324,6 +324,21 @@ impl BreezServices {
         self.persister.get_node_state()
     }
 
+    /// Sign given message with privkey
+    pub async fn sign_message(&self, message: String) -> Result<String> {
```
Comment (roeierez) [inline libs/sdk-core/src/breez_services.rs:328]: Recently we started shifting towards using structs for request response, e.g:
SignMessageRequest
SignMessageResponse
It makes it easier to maintain backward compatibility. Could you please use this approach?

**Review by roeierez — CHANGES_REQUESTED** [review]
Thank you @lnd3v looks good, dropped some comments.

File: libs/sdk-react-native/src/index.tsx [hunk 2]
```diff
@@ -515,6 +515,14 @@ export const connect = async (config: Config, seed: Uint8Array): Promise<void> =
     await BreezSDK.connect(config, seed)
 }
 
+export const sign_message = async (message: string): Promise<string> => {
+    return await BreezSDK.sign_message(message)
+}
+
+export const check_message = async (message: string, pubkey: string, zbase: string): Promise<boolean> => {
+    return await BreezSDK.check_message(message, pubkey, zbase)
```
Comment (dangeross) [inline libs/sdk-react-native/src/index.tsx:523]: These function names should be camel-cased.

**Review by dangeross — CHANGES_REQUESTED** [review]

File: libs/sdk-core/src/greenlight/node_api.rs [hunk 3]
```diff
@@ -518,6 +518,28 @@ impl NodeAPI for Greenlight {
...
+    async fn sign_message(&self, message: &str) -> Result<String> {
+        let mut client = self.get_node_client().await?;
+        let request = pb::cln::SignmessageRequest {
+            message: message.to_string(),
+        };
+        let response = client.sign_message(request).await?.into_inner();
+        Ok(response.zbase)
+    }
+
+    async fn check_message(&self, message: &str, pubkey: &str, zbase: &str) -> Result<bool> {
```
[the remark this answers is not in the document]
Comment (lnd3v) [reply libs/sdk-core/src/greenlight/node_api.rs:530]: Sure, that's done.

**Review by roeierez — CHANGES_REQUESTED** [review]

**Comment by lnd3v** [comment]
For the record, i had to ask about signing on the greenlight discord. The response I got:

```
cdecker — Today at 15:50
We are currently using the signmessage method to create attestations and authorizations for keyless clients. If we were to allow clients to call signmessage they could escalate privileges trivially.

We are going to change that soon and signmessage will be enabled. In the meantime you may find the vls code helpful in deriving the secret and signing the message.
```

After that it was easily to find the signing code. It works when doing a copy and paste, however when trying to use the vls code directly it fails as trait fields are private. If they were public it would be as simple as:

```
# cargo.toml
vls-core = "0.9.1-rc.2"
# node_api.rs
use lightning_signer::signer::derive::{KeyDerive, NativeKeyDerive};
    async fn sign_message(&self, message: &str) -> Result<String> {
        let native_key_derive = NativeKeyDerive { network: bitcoin::network::constants::Network::Bitcoin };
        let (_pubkey_from_node_id, seckey) = KeyDerive::node_keys(&native_key_derive, &self.secret, &Secp256k1::new());
        println!("_pubkey_from_node_id: {} ", hex::encode(_pubkey_from_node_id.serialize()));
        Ok(sign(message.as_bytes(), &seckey).unwrap())
    }
```
However since the NativeKeyDerive trait doesn't have a public constructor i'll go ahead and do the copy & paste approach for now, which will just have to get cleaned out once the signer's sign_message works as expected for signing lightning messages.

File: libs/sdk-react-native/src/index.tsx [hunk 4]
```diff
@@ -515,6 +534,14 @@ export const connect = async (config: Config, seed: Uint8Array): Promise<void> =
     await BreezSDK.connect(config, seed)
 }
 
+export const signMessage = async (request: SignMessageRequest): Promise<SignMessageResponse> => {
```
Comment (lnd3v) [inline libs/sdk-react-native/src/index.tsx:537]: I'm not clear whether the signMessage / checkMessage functions should take string arguments for the Request objects.

File: libs/sdk-core/Cargo.toml [hunk 5]
```diff
@@ -58,6 +58,8 @@ tempfile = "3"
 thiserror = "*"
 const_format = "*"
 miniz_oxide = "0.7.1"
+secp256k1 = "0.24.3"
```
Comment (JssDWt) [inline libs/sdk-core/Cargo.toml:61]: Let's remove this dependency

File: libs/sdk-core/src/greenlight/node_api.rs [hunk 6]
```diff
@@ -1,42 +1,48 @@
...
 use tonic::Streaming;
 
-use crate::invoice::parse_invoice;
-use crate::models::{
-    Config, GreenlightCredentials, LnPaymentDetails, Network, NodeAPI, NodeState, PaymentDetails,
-    PaymentType, SyncResponse, UnspentTransactionOutput,
-};
-use crate::persist::db::SqliteStorage;
-use crate::{Channel, ChannelState, NodeConfig};
-
```
Comment (JssDWt) [inline libs/sdk-core/src/greenlight/node_api.rs:39]: I think only `use bitcoin::secp256k1::PublicKey;` and `use lightning::util::message_signing;` are necessary here. The other changes can be dropped. Also the zbase32 is not necessary, because it's referenced directly in the file.

File: libs/sdk-core/src/greenlight/node_api.rs [hunk 7]
```diff
@@ -62,13 +68,12 @@ impl Greenlight {
...
-        let encryption_key = Self::derive_bip32_key(
+        let extended = Self::derive_bip32_key(
             config.network,
             &signer,
             vec![ChildNumber::from_hardened_idx(140)?, ChildNumber::from(0)],
-        )?
-        .to_priv()
-        .to_bytes();
+        )?;
+        let encryption_key = extended.to_priv().to_bytes();
```
Comment (JssDWt) [inline libs/sdk-core/src/greenlight/node_api.rs:76]: nit: Unrelated change?

File: libs/sdk-core/src/greenlight/node_api.rs [hunk 8]
```diff
@@ -337,12 +343,13 @@ impl NodeAPI for Greenlight {
...
-                amount_to_msat(&parse_amount(c.spendable.clone()).unwrap_or_default())
-            })
-            .sum::<u64>();
+        let channels_balance = offchain_funds.iter().fold(0, |a, b| {
+            let hex_txid = hex::encode(b.funding_txid.clone());
+            if opened_channels.iter().any(|c| c.funding_txid == hex_txid) {
+                return a + b.our_amount_msat;
+            }
+            a
+        });
```
Comment (JssDWt) [inline libs/sdk-core/src/greenlight/node_api.rs:352]: nit: Unrelated change?

File: libs/sdk-core/src/greenlight/node_api.rs [hunk 9]
```diff
@@ -972,11 +991,10 @@ impl TryFrom<ListclosedchannelsClosedchannels> for crate::models::Channel {
 
 #[cfg(test)]
 mod tests {
+    use crate::models;
     use anyhow::Result;
     use gl_client::pb;
 
-    use crate::models;
-
```
Comment (JssDWt) [inline libs/sdk-core/src/greenlight/node_api.rs:979]: nit: Unrelated change?

**Review by JssDWt — CHANGES_REQUESTED** [review]
Some small comments about cleaning up.

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt [hunk 10]
```diff
@@ -82,6 +82,30 @@ fun asLnUrlWithdrawRequestData(reqData: ReadableMap): LnUrlWithdrawRequestData?
     return null
 }
 
+fun asCheckMessageRequest(reqData: ReadableMap): CheckMessageRequest? {
+    val callback = reqData.getString("callback")
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt:86]: This callback variable can be removed

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt [hunk 11]
```diff
@@ -82,6 +82,30 @@ fun asLnUrlWithdrawRequestData(reqData: ReadableMap): LnUrlWithdrawRequestData?
...
+
+    if (callback != null && message != null && signature != null && pubkey != null) {
+        return CheckMessageRequest(message, signature, pubkey)
+    }
+
+    return null
+}
+
+fun asSignMessageRequest(reqData: ReadableMap): SignMessageRequest? {
+    val callback = reqData.getString("callback")
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKMapper.kt:99]: This callback variable can be removed

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt [hunk 12]
```diff
@@ -13,6 +13,24 @@ import java.util.concurrent.ExecutorService
...
+data class CheckMessageRequest(
+    val message: String,
+    val pubkey: String,
+    val signature: String
+)
+
+data class CheckMessageResponse(
+    val is_valid: Boolean
+)
+
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt:33]: These can be removed as they will be generated from the udl

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt [hunk 13]
```diff
@@ -146,6 +164,46 @@ class BreezSDKModule(reactContext: ReactApplicationContext) : ReactContextBaseJa
...
 
+    @ReactMethod
+    fun signMessage(reqData: ReadableMap, message: String, promise: Promise) {
+        executor.execute {
+            val signMessageRequest = asSignMessageRequest(reqData)
+            if signMessageRequest == null {
+                 promise.reject(TAG, "Invalid reqData")
+            } else {
+                try {
+                    val response = getBreezServices().sign_message(signMessageRequest)
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt:175]: sign_message should be signMessage

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt [hunk 14]
```diff
@@ -146,6 +164,46 @@ class BreezSDKModule(reactContext: ReactApplicationContext) : ReactContextBaseJa
         }
     }
 
+    @ReactMethod
+    fun signMessage(reqData: ReadableMap, message: String, promise: Promise) {
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt:168]: message can be removed

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt [hunk 15]
```diff
@@ -146,6 +164,46 @@ class BreezSDKModule(reactContext: ReactApplicationContext) : ReactContextBaseJa
...
+                } catch (e: SdkException) {
+                    e.printStackTrace()
+                    promise.reject(TAG, e.message ?: "Error calling signMessage", e)
+                }
+            }
+        }
+    }
+
+    @ReactMethod
+    fun checkMessage(reqData: ReadableMap, message: String, signature: String, pubkey: String, promise: Promise) {
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt:187]: message, signature & pubkey can be removed

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt [hunk 16]
```diff
@@ -146,6 +164,46 @@ class BreezSDKModule(reactContext: ReactApplicationContext) : ReactContextBaseJa
...
+        executor.execute {
+            val checkMessageRequest = asCheckMessageRequest(reqData)
+            if checkMessageRequest == null {
+                 promise.reject(TAG, "Invalid reqData")
+            } else {
+                try {
+                    val response = getBreezServices().check_message(checkMessageRequest)
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt:194]: check_message should be checkMessage

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt [hunk 17]
```diff
@@ -146,6 +164,46 @@ class BreezSDKModule(reactContext: ReactApplicationContext) : ReactContextBaseJa
...
+    fun signMessage(reqData: ReadableMap, message: String, promise: Promise) {
+        executor.execute {
+            val signMessageRequest = asSignMessageRequest(reqData)
+            if signMessageRequest == null {
+                 promise.reject(TAG, "Invalid reqData")
+            } else {
+                try {
+                    val response = getBreezServices().sign_message(signMessageRequest)
+                    val signedMessage = response.signed_message
+                    promise.resolve(signedMessage)
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt:177]: The SignMessageResponse object should be returned here, you can add a mapping function to `BreezSDKMapper.kt`

```
fun readableMapOf(signMessageResponse: SignMessageResponse): ReadableMap {
    return readableMapOf(
            "signedMessage" to signMessageResponse.signedMessage
    )
}
```

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt [hunk 18]
```diff
@@ -146,6 +164,46 @@ class BreezSDKModule(reactContext: ReactApplicationContext) : ReactContextBaseJa
...
+    fun checkMessage(reqData: ReadableMap, message: String, signature: String, pubkey: String, promise: Promise) {
+        executor.execute {
+            val checkMessageRequest = asCheckMessageRequest(reqData)
+            if checkMessageRequest == null {
+                 promise.reject(TAG, "Invalid reqData")
+            } else {
+                try {
+                    val response = getBreezServices().check_message(checkMessageRequest)
+                    val signedMessage = response.signed_message
+                    promise.resolve(signedMessage)
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt:196]: The CheckMessageResponse object should be returned here, you can add a mapping function to `BreezSDKMapper.kt`

```
fun readableMapOf(checkMessageResponse: CheckMessageResponse): ReadableMap {
    return readableMapOf(
            "isValid" to checkMessageResponse.isValid
    )
}
```

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt [hunk 19]
```diff
@@ -146,6 +164,46 @@ class BreezSDKModule(reactContext: ReactApplicationContext) : ReactContextBaseJa
...
+                } catch (e: SdkException) {
+                    e.printStackTrace()
+                    promise.reject(TAG, e.message ?: "Error calling signMessage", e)
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt:199]: Can be changed to
```
promise.reject(e.javaClass.simpleName, e.message, e)
```

File: libs/sdk-react-native/ios/RNBreezSDK.swift [hunk 20]
```diff
@@ -1,6 +1,24 @@
...
+struct CheckMessageRequest {
+    var message: String
+    var pubkey: String
+    var signature: String
+}
+
+struct CheckMessageResponse {
+    var isValid: Bool
+}
+
```
Comment (dangeross) [inline libs/sdk-react-native/ios/RNBreezSDK.swift:21]: These can be removed as they will be generated from the udl

File: libs/sdk-react-native/ios/RNBreezSDK.swift [hunk 21]
```diff
@@ -111,6 +129,26 @@ class RNBreezSDK: RCTEventEmitter {
             reject(RNBreezSDK.TAG, "Invalid config", nil)
         }
     }
+
+    @objc(signMessage:resolver:rejecter:)
```
Comment (dangeross) [inline libs/sdk-react-native/ios/RNBreezSDK.swift:133]: This also needs adding into `RNBreezSDK.m`

```
RCT_EXTERN_METHOD(
    signMessage: (NSDictionary*)reqData
    resolver: (RCTPromiseResolveBlock)resolve
    rejecter: (RCTPromiseRejectBlock)reject
)
```

File: libs/sdk-react-native/ios/RNBreezSDK.swift [hunk 22]
```diff
@@ -111,6 +129,26 @@ class RNBreezSDK: RCTEventEmitter {
...
+    func signMessage(_ request: SignMessageRequest, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) -> SignMessageResponse {
+        do {
+            let response = try BreezSDK.sign_message(message)
+            resolve(response)
+        } catch let error {
+            reject(RNBreezSDK.TAG, "Error calling signMessage", error)
+        }
+    }
+
+    @objc(checkMessage:pubkey:signature:resolver:rejecter:)
```
Comment (dangeross) [inline libs/sdk-react-native/ios/RNBreezSDK.swift:143]: This also needs adding into `RNBreezSDK.m`

```
RCT_EXTERN_METHOD(
    checkMessage: (NSDictionary*)reqData
    resolver: (RCTPromiseResolveBlock)resolve
    rejecter: (RCTPromiseRejectBlock)reject
)
```

Comment (dangeross) [inline libs/sdk-react-native/ios/RNBreezSDK.swift:143]: pubkey and signature can be removed

File: libs/sdk-react-native/ios/RNBreezSDK.swift [hunk 23]
```diff
@@ -111,6 +129,26 @@ class RNBreezSDK: RCTEventEmitter {
...
     }
+
+    @objc(signMessage:resolver:rejecter:)
+    func signMessage(_ request: SignMessageRequest, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) -> SignMessageResponse {
+        do {
+            let response = try BreezSDK.sign_message(message)
+            resolve(response)
+        } catch let error {
+            reject(RNBreezSDK.TAG, "Error calling signMessage", error)
+        }
```
Comment (dangeross) [inline libs/sdk-react-native/ios/RNBreezSDK.swift:140]: This function should convert the reqData into a SignMessageRequest struct, then return a dictionary mapping the SignMessageResponse

```
        if let signMessageRequest = BreezSDKMapper.asSignMessageRequest(reqData: reqData) {
            do {
                let signMessageResponse = try getBreezServices().signMessage(reqData: signMessageRequest)                
                resolve(BreezSDKMapper.dictionaryOf(signMessageResponse: signMessageResponse))
            } catch let err {
                rejectErr(err: err, reject: reject)      
            }
        } else {            
            rejectErr(err: SdkError.Generic(message:"Invalid reqData"), reject: reject)
        }
```

File: libs/sdk-react-native/ios/RNBreezSDK.swift [hunk 24]
```diff
@@ -111,6 +129,26 @@ class RNBreezSDK: RCTEventEmitter {
...
+    }
+
+    @objc(checkMessage:pubkey:signature:resolver:rejecter:)
+    func checkMessage(_ request: CheckMessageRequest, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) -> CheckMessageResponse {
+        do {
+            let response = try BreezSDK.check_message(request)
+            resolve(response)
+        } catch let error {
+            reject(RNBreezSDK.TAG, "Error calling checkMessage", error)
+        }
```
Comment (dangeross) [inline libs/sdk-react-native/ios/RNBreezSDK.swift:150]: This function should convert the reqData into a CheckMessageRequest struct, then return a dictionary mapping the CheckMessageResponse

```
        if let checkMessageRequest = BreezSDKMapper.asCheckMessageRequest(reqData: reqData) {
            do {
                let checkMessageResponse = try getBreezServices().checkMessage(reqData: checkMessageRequest)                
                resolve(BreezSDKMapper.dictionaryOf(checkMessageResponse: checkMessageResponse))
            } catch let err {
                rejectErr(err: err, reject: reject)      
            }
        } else {            
            rejectErr(err: SdkError.Generic(message:"Invalid reqData"), reject: reject)
        }
```

File: libs/sdk-react-native/ios/RNBreezSDK.swift [hunk 25]
```diff
@@ -111,6 +129,26 @@ class RNBreezSDK: RCTEventEmitter {
             reject(RNBreezSDK.TAG, "Invalid config", nil)
         }
     }
+
+    @objc(signMessage:resolver:rejecter:)
+    func signMessage(_ request: SignMessageRequest, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) -> SignMessageResponse {
```
Comment (dangeross) [inline libs/sdk-react-native/ios/RNBreezSDK.swift:134]: This reqData type should be `reqData:[String: Any]` as it otherwise causes issues being transferred over the RN bridge

File: libs/sdk-react-native/ios/RNBreezSDK.swift [hunk 26]
```diff
@@ -111,6 +129,26 @@ class RNBreezSDK: RCTEventEmitter {
...
+        do {
+            let response = try BreezSDK.sign_message(message)
+            resolve(response)
+        } catch let error {
+            reject(RNBreezSDK.TAG, "Error calling signMessage", error)
+        }
+    }
+
+    @objc(checkMessage:pubkey:signature:resolver:rejecter:)
+    func checkMessage(_ request: CheckMessageRequest, resolver resolve: @escaping RCTPromiseResolveBlock, rejecter reject: @escaping RCTPromiseRejectBlock) -> CheckMessageResponse {
```
Comment (dangeross) [inline libs/sdk-react-native/ios/RNBreezSDK.swift:144]: This reqData type should be `reqData:[String: Any]` as it otherwise causes issues being transferred over the RN bridge

File: libs/sdk-react-native/src/index.tsx [hunk 27]
```diff
@@ -515,6 +534,14 @@ export const connect = async (config: Config, seed: Uint8Array): Promise<void> =
...
+    return await BreezSDK.sign_message(request);
```
Comment (dangeross) [inline libs/sdk-react-native/src/index.tsx:538]: Needs to be camel cased, as the native function is

File: libs/sdk-react-native/src/index.tsx [hunk 28]
```diff
@@ -515,6 +534,14 @@ export const connect = async (config: Config, seed: Uint8Array): Promise<void> =
...
+}
+
+export const checkMessage = async (request: CheckMessageRequest): Promise<CheckMessageResponse> => {
+    return await BreezSDK.check_message(request);
```
Comment (dangeross) [inline libs/sdk-react-native/src/index.tsx:542]: Needs to be camel cased, as the native function is

File: libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt [hunk 29]
```diff
@@ -146,6 +164,46 @@ class BreezSDKModule(reactContext: ReactApplicationContext) : ReactContextBaseJa
...
+            if signMessageRequest == null {
+                 promise.reject(TAG, "Invalid reqData")
+            } else {
+                try {
+                    val response = getBreezServices().sign_message(signMessageRequest)
+                    val signedMessage = response.signed_message
+                    promise.resolve(signedMessage)
+                } catch (e: SdkException) {
+                    e.printStackTrace()
+                    promise.reject(TAG, e.message ?: "Error calling signMessage", e)
```
Comment (dangeross) [inline libs/sdk-react-native/android/src/main/java/com/breezsdk/BreezSDKModule.kt:180]: Can be changed to
```
promise.reject(e.javaClass.simpleName, e.message, e)
```

**Review by dangeross — CHANGES_REQUESTED** [review]

**Comment by JssDWt** [comment]
- Rebased
- Upgraded greenlight dependency to include the merged sign_message PR
- Fixed react native bindings. Android works, iOS needs to be tested still.

**Comment by JssDWt** [comment]
Renamed `signed_message` to `signature`

**Comment by JssDWt** [comment]
Added sign/check message to the flutter bridge 
@erdemyerebasmaz can you check whether I did that right?

**Comment by JssDWt** [comment]
- rebased
- Updated swift mappings in sdk-react-native. iOS now also works.

File: libs/sdk-react-native/ios/BreezSDKMapper.swift [hunk 30]
```diff
@@ -201,6 +219,12 @@ class BreezSDKMapper {
             "message": bitcoinAddressData.message,
         ]
     }
+
+    static func dictionaryOf(checkMessageResponse: CheckMessageResponse) -> [String: Any?] {
```
Comment (dangeross) [inline libs/sdk-react-native/ios/BreezSDKMapper.swift:223]: `checkMessageResponse.isValid` should always return a value so the return type can be `[String: Any]`

File: libs/sdk-react-native/ios/BreezSDKMapper.swift [hunk 31]
```diff
@@ -508,6 +532,12 @@ class BreezSDKMapper {
             "htlcMaximumMsat": routeHintHop.htlcMaximumMsat
         ]
     }
+
+    static func dictionaryOf(signMessageResponse: SignMessageResponse) -> [String: Any?] {
```
Comment (dangeross) [inline libs/sdk-react-native/ios/BreezSDKMapper.swift:536]: `signMessageResponse.signature` should always return a value so the return type can be `[String: Any]`

**Review by dangeross — APPROVED** [review]
Looks good! Left 2 nitty comments

**Comment by erdemyerebasmaz** [comment]
> Added sign/check message to the flutter bridge @erdemyerebasmaz can you check whether I did that right?

Looks good!

**Comment by erdemyerebasmaz** [comment]
> Added sign/check message to the flutter bridge @erdemyerebasmaz can you check whether I did that right?

@JssDWt looks good!

**Review by JssDWt — APPROVED** [review]

**Comment by JssDWt** [comment]
PR is ready to merge.
Is it a problem that the greenlight dependency is updated in this PR @ok300 @roeierez?

File: libs/sdk-core/src/breez_services.rs [hunk 32]
```diff
@@ -96,6 +96,34 @@ pub struct InvoicePaidDetails {
...
+    /// The signature that covers the message of SignMessageRequest
+    pub signature: String,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct CheckMessageRequest {
+    /// The message to be checked
+    pub message: String,
+    /// The public key
+    pub pubkey: String,
```
Comment (ok300) [inline libs/sdk-core/src/breez_services.rs:116]: Can a node verify signatures for multiple pubkeys? (multiple derivation paths?)

If not, isn't the pubkey here superfluous?

**Review by ok300 — COMMENTED** [review]
Generally LGTM, left a few comments.

I don't think it's a problem that the GL dependencies were updated, especially since the new GL version is needed for this feature.

File: libs/sdk-core/src/binding.rs [hunk 33]
```diff
@@ -89,6 +90,16 @@ pub fn disconnect() -> Result<()> {
     })
 }
 
+/// Sign given message with privkey
```
[the remark this answers is not in the document]
Comment (JssDWt) [reply libs/sdk-core/src/binding.rs:93]: It's the node id of your lightning node. Will update the comment to reflect this.

File: libs/sdk-core/src/breez_services.rs (hunk 32 above)
Comment (JssDWt) [reply libs/sdk-core/src/breez_services.rs:116]: It verifies that the given signature is a signature over the specified message, signed by the private key belonging to the passed pubkey.
So if another lightning node signs a message, you can verify it was that other lightning node signing the message. If it was not this pubkey that signed the message, the signature is invalid.
It's implemented this way in all lightning implementations, so this is cross compatible with cln, lnd, etc.

Comment (ok300) [reply libs/sdk-core/src/breez_services.rs:116]: Oh ok, so `check_message` can check any message signed by any node, not just the own node.

**Comment by ok300** [comment]
Fixes #109

File: libs/sdk-core/src/breez_services.rs (hunk 32 above)
Comment (JssDWt) [reply libs/sdk-core/src/breez_services.rs:116]: Yeah exactly

File: libs/sdk-core/src/binding.rs [hunk 34]
```diff
@@ -89,6 +90,16 @@ pub fn disconnect() -> Result<()> {
...
+pub fn sign_message(request: SignMessageRequest) -> Result<SignMessageResponse> {
+    block_on(async { get_breez_services()?.sign_message(request).await })
+}
+
+/// Check given message for pubkey with signature (zbase)
+pub fn check_message(request: CheckMessageRequest) -> Result<CheckMessageResponse> {
```
[the remark this answers is not in the document]
Comment (JssDWt) [reply libs/sdk-core/src/binding.rs:99]: @ok300 I think the result indicates error/no error, and the boolean indicates whether the signature is valid or not valid. So the response should have a value, and then this comment applies: https://github.com/breez/breez-sdk/pull/319#discussion_r1263431401 ?

**Comment by JssDWt** [comment]
- Updated documentation on functions and structs.
- Rebased

File: libs/sdk-core/src/binding.rs (hunk 34 above)
[the remark this answers is not in the document]
Comment (ok300) [reply libs/sdk-core/src/binding.rs:99]: Either way is fine with me.

In other similar places where we run validation, the method returns a `Result<()>`. If it's `Ok`, validation passed, if `Err`, which can contain a custom error message, it means validation failed.

[the remark this answers is not in the document]
Comment (ok300) [reply libs/sdk-core/src/binding.rs:99]: As to the struct comment linked above, I agree, when we return values, a nicely-named struct for Request / Response is better.

IMO you can leave this as it is, no need to change it.
